File: TypeScript/nestshop/src/admin-settings/admin-settings.service.ts
import { Injectable } from '@nestjs/common';
import { InjectRepository } from '@nestjs/typeorm';
import { Repository } from 'typeorm';
import { CacheService } from '../common/cache/cache.service';
import { SetAllowedExtensionsDto } from './dto/set-allowed-extensions.dto';
import { UpdateReviewPolicyDto } from './dto/update-review-policy.dto';
import { UpdateUploadLimitsDto } from './dto/update-upload-limits.dto';
import { SystemSetting } from './entities/system-setting.entity';

@Injectable()
export class AdminSettingsService {
  private readonly defaultExtensions = ['jpg', 'jpeg', 'png', 'webp', 'gif', 'mp4', 'mp3', 'pdf'];
  private readonly defaultUploadLimits = { image: 5, video: 100, audio: 20 };
  private readonly defaultReviewPolicy = { maxImageCount: 10, pointAmount: 500 };

  constructor(
    @InjectRepository(SystemSetting)
    private systemSettingRepository: Repository<SystemSetting>,
    private readonly cacheService: CacheService,
  ) {}

  async getAllowedExtensions() {
    const setting = await this.getOrCreate('allowed_extensions', { extensions: this.defaultExtensions });
    return { extensions: this.normalizeExtensions(setting.settingValue.extensions as string[]) };
  }

  async setAllowedExtensions(dto: SetAllowedExtensionsDto) {
    const normalized = this.normalizeExtensions(dto.extensions);
    const setting = await this.upsert('allowed_extensions', { extensions: normalized });
    await this.cacheService.del('settings:allowed_extensions');
    return { extensions: this.normalizeExtensions(setting.settingValue.extensions as string[]) };
  }

  async getUploadLimits() {
    const setting = await this.getOrCreate('upload_limits', this.defaultUploadLimits);
    const value = setting.settingValue;

    return {
      image: Number(value.image ?? this.defaultUploadLimits.image),
      video: Number(value.video ?? this.defaultUploadLimits.video),
      audio: Number(value.audio ?? this.defaultUploadLimits.audio),
    };
  }

  async updateUploadLimits(dto: UpdateUploadLimitsDto) {
    const current = await this.getUploadLimits();

    const merged = {
      image: dto.image ?? current.image,
      video: dto.video ?? current.video,
      audio: dto.audio ?? current.audio,
    };

    const setting = await this.upsert('upload_limits', merged);
    const value = setting.settingValue;

    return {
      image: Number(value.image),
      video: Number(value.video),
      audio: Number(value.audio),
    };
  }

  async getReviewPolicy() {
    const setting = await this.getOrCreate('review_policy', this.defaultReviewPolicy);
    const value = setting.settingValue;

    return {
      maxImageCount: Number(value.maxImageCount ?? this.defaultReviewPolicy.maxImageCount),
      pointAmount: Number(value.pointAmount ?? this.defaultReviewPolicy.pointAmount),
    };
  }

  async updateReviewPolicy(dto: UpdateReviewPolicyDto) {
    const setting = await this.upsert('review_policy', {
      maxImageCount: dto.maxImageCount,
      pointAmount: dto.pointAmount,
    });

    const value = setting.settingValue;
    return {
      maxImageCount: Number(value.maxImageCount),
      pointAmount: Number(value.pointAmount),
    };
  }

  private async getOrCreate(key: string, defaultValue: Record<string, unknown>) {
    const existing = await this.systemSettingRepository.findOne({ where: { settingKey: key } });
    if (existing) {
      return existing;
    }

    const created = this.systemSettingRepository.create({
      settingKey: key,
      settingValue: defaultValue,
    });

    return this.systemSettingRepository.save(created);
  }

  private async upsert(key: string, value: Record<string, unknown>) {
    let existing = await this.systemSettingRepository.findOne({ where: { settingKey: key } });

    if (!existing) {
      existing = this.systemSettingRepository.create({ settingKey: key, settingValue: value });
    } else {
      existing.settingValue = value;
    }

    return this.systemSettingRepository.save(existing);
  }

  private normalizeExtensions(extensions: string[]) {
    const normalized = extensions.map((ext) => ext.trim().toLowerCase()).filter(Boolean);
    return [...new Set(normalized)].sort();
  }
}

File: BackEnd/Kotlin/kotlin-ktor-gradle-exposeddao-postgresql/src/main/kotlin/com/pbshop/kotlin/ktor/gradle/exposeddao/postgresql/adminsettings/AdminSettingsService.kt
package com.pbshop.kotlin.ktor.gradle.exposeddao.postgresql.adminsettings

import com.pbshop.kotlin.ktor.gradle.exposeddao.postgresql.common.PbShopException
import com.pbshop.kotlin.ktor.gradle.exposeddao.postgresql.common.StubResponse
import io.ktor.http.HttpStatusCode

class AdminSettingsService(
    private val repository: AdminSettingsRepository,
) {
    fun extensions(): StubResponse = StubResponse(data = repository.getAllowedExtensions())

    fun setExtensions(request: SetAllowedExtensionsRequest): StubResponse {
        val normalized =
            request.extensions
                .map { it.trim().lowercase() }
                .filter { it.isNotBlank() }
                .distinct()
                .sorted()
        if (normalized.isEmpty()) {
            throw PbShopException(HttpStatusCode.BadRequest, "VALIDATION_ERROR", "extensions가 필요합니다.")
        }
        return StubResponse(data = repository.setAllowedExtensions(normalized))
    }

    fun uploadLimits(): StubResponse = StubResponse(data = repository.getUploadLimits())

    fun updateUploadLimits(request: UpdateUploadLimitsRequest): StubResponse {
        if (listOf(request.image, request.video, request.audio).all { it == null }) {
            throw PbShopException(HttpStatusCode.BadRequest, "VALIDATION_ERROR", "수정할 용량 제한 값이 필요합니다.")
        }
        listOf(request.image, request.video, request.audio).filterNotNull().forEach {
            if (it <= 0) {
                throw PbShopException(HttpStatusCode.BadRequest, "VALIDATION_ERROR", "용량 제한은 1MB 이상이어야 합니다.")
            }
        }
        return StubResponse(
            data =
                repository.updateUploadLimits(
                    image = request.image,
                    video = request.video,
                    audio = request.audio,
                ),
        )
    }

    fun reviewPolicy(): StubResponse = StubResponse(data = repository.getReviewPolicy())

    fun updateReviewPolicy(request: UpdateReviewPolicyRequest): StubResponse {
        if (request.maxImageCount <= 0 || request.pointAmount < 0) {
            throw PbShopException(HttpStatusCode.BadRequest, "VALIDATION_ERROR", "유효한 리뷰 정책 값이 필요합니다.")
        }
        return StubResponse(
            data =
                repository.updateReviewPolicy(
                    maxImageCount = request.maxImageCount,
                    pointAmount = request.pointAmount,
                ),
        )
    }
}
